back-end/service.py

```diff
@@ -1,172 +1,301 @@
-    key_size=2048
-	)
-	private_key = key.private_bytes(
-	crypto_serialization.Encoding.PEM,
-    crypto_serialization.PrivateFormat.PKCS8,
-    crypto_serialization.NoEncryption())
-	public_key = key.public_key().public_bytes(
-    crypto_serialization.Encoding.OpenSSH,
-    crypto_serialization.PublicFormat.OpenSSH
-	)
-
-	SSH_PATH = "../ssh/" + user
-	os.makedirs(SSH_PATH, exist_ok=True)
-
-	with open(SSH_PATH + '/gin_id_rsa', 'w+') as private_key_file: 
-		private_key_file.write(private_key.decode('utf-8'))
-	with open(SSH_PATH + '/gin_id_rsa.pub', 'w+') as public_key_file: 
-		public_key_file.write(public_key.decode('utf-8'))
-	os.chmod(SSH_PATH, 0o700)
-	os.chmod(SSH_PATH + '/gin_id_rsa', 0o600)
-	os.chmod(SSH_PATH + '/gin_id_rsa.pub', 0o600)
-	
-	#os.environ['GIT_SSH_COMMAND'] = "ssh -i " + SSH_PATH + "/gin_id_rsa"
-
-	response = requests.post(path + "/api/v1/user/keys", 
-	headers={'Authorization': 'token ' + str(token.token)},
-	data={'title': 'gin_id_rsa', 'key': public_key}
-	)
-
-	return 'gin_id_rsa'
-
-def validUser(auth):
-	return api.valid_authentication(auth)
+
+
+def user(token):
+    response = requests.get(
+        path + "/api/v1/user",
+        headers={'Authorization': 'token ' + str(token)}
+        ).json()
+
+    return response
+
+
+def ensureToken(username, password):
+
+    res = requests.get(
+        path + "/api/v1/users/{}/tokens".format(username),
+        auth=(username, password)).json()
+
+    for token in res:
+        if token['name'] == 'gin-proc':
+            return token['sha1']
+
+    res = requests.post(
+        path + "/api/v1/users/{}/tokens".format(username),
+        auth=(username, password),
+        data={'name': 'gin-proc'}
+    ).json()
+
+    return res['sha1'] 
+
+
+def ensureKeysOnServer(token):
+
+    response = requests.get(
+        path + "/api/v1/user/keys",
+        headers={'Authorization': 'token ' + str(token.token)}
+        )
+
+    for keys in response.json():
+        if keys['title'] == 'gin_id_rsa':
+
+            print('Key {} installed on the server'.format('gin_id_rsa'))
+            return (True)
+
+
+def ensureKeysOnLocal(path):
+
+    KEY_PATH = os.path.join(path, 'gin_id_rsa')
+    EXISTS = os.path.exists(KEY_PATH)
+    print('Key {} installed locally'.format('gin_id_rsa') is EXISTS)
+    return EXISTS
+
+
+def installFreshKeys(SSH_PATH, token):
+
+        key = rsa.generate_private_key(
+            backend=default_backend(),
+            public_exponent=65537,
+            key_size=2048
+        )
+        private_key = key.private_bytes(
+            serialization.Encoding.PEM,
+            serialization.PrivateFormat.PKCS8,
+            serialization.NoEncryption()
+            )
+        public_key = key.public_key().public_bytes(
+            serialization.Encoding.OpenSSH,
+            serialization.PublicFormat.OpenSSH
+        )
+
+        os.makedirs(SSH_PATH, exist_ok=True)
+
+        with open(
+            os.path.join(SSH_PATH, '/gin_id_rsa'),
```

Comment:
This is probably why the whole key business wasn't working.  When one of the components in `os.path.join()` starts with a `/`, the previous parts are ignored, so this line produces `/gin_id_rsa`.

```suggestion
            os.path.join(SSH_PATH, 'gin_id_rsa'),
```